src/java.base/share/classes/java/nio/file/Path.java

```diff
@@ -248,6 +250,151 @@ public static Path of(URI uri) {
+     * or a default value.  The extension is defined to be the portion of
+     * the {@code String} representation of the file name after the last
+     * dot ('.').  If the first character in the file name string is a dot
+     * it is ignored.  If the extension cannot be determined, then the
+     * parameter {@code defaultExtension} is returned.  This will occur if
+     * the path has zero elements ({@link #getFileName()} returns
+     * {@code null}), or the file name string does not contain a dot, only
+     * the first character is a dot, or the last character is a dot.
+     *
+     * @implSpec
+     * The default implementation is equivalent for this path to:
+     * <pre>{@code
+     *     String name = getFileName().toString();
+     *     int lastDot = name.lastIndexOf('.');
+     *     lastDot > 0 && lastDot < name.length() - 1 ?
+     *         name.substring(lastDot + 1) : defaultExtension;
+     * }</pre>
+     *
+     * @param   defaultExtension
+     *          the value to return if the extension is indeterminate;
+     *          may be {@code null}
+     *
+     * @return  the file name extension of this path, or
+     *          {@code defaultExtension} if the extension is indeterminate
+     *
+     * @since 19
+     */
+    default String getExtension(String defaultExtension) {
+        Path fileName = getFileName();
+        if (fileName == null) {
+            return defaultExtension;
+        }
+
+        String fileNameString = fileName.toString();
+        int length = fileNameString.length();
+
+        // Indeterminate if fileNameString is too short
+        if (length > 2) {
+            int lastDotIndex = fileNameString.lastIndexOf('.');
+            // Indeterminate if no dot or found at last or only the first index
+            if (lastDotIndex > 0 && lastDotIndex < length - 1) {
+                return fileNameString.substring(lastDotIndex + 1);
+            }
+        }
+
+        return defaultExtension;
+    }
+
+    /**
+     * Returns whether the string representation of this path has an
+     * extension among the one or more extensions given as parameters.
+     * The extension of this path will be compared against the given
+     * parameters in the order in which they are supplied.
+     *
+     * @param   ext
+     *          the first extension for which to check
+     *
+     * @param   extensions
+     *          zero or more subsequent extensions for which to check
+     *
+     * @return  an {@code Optional} which either contains the found extension
+     *          or is empty if this path either does not have an extension or
+     *          its extension is not among those given
+     *
+     * @since 19
+     */
+    default Optional<String> hasExtension(String ext, String... extensions) {
+        String thisExtension = getExtension(null);
+        if (thisExtension == null)
+            return Optional.empty();
+
+        Objects.requireNonNull(ext);
```

Comment:
Typically, all of the argument checks are done first. 
Though it does create a strange shape for the code when there is no extension. Two passes over the extensions seems like a waste.

Reply:
Yes, I thought about this. It did not seem to be worthwhile to pass through the args just to check for a null parameter that one might not even need to use.